File: bar2.py
import OpenGL.GLUT as glut
import OpenGL.GL as gl
import numpy as np

class FruitData:
    def __init__(self, name, quantity):
        self.name = name
        self.quantity = quantity

fruitData = [
    FruitData("Avocados", 36),
    FruitData("Oranges", 41),
    FruitData("Bananas", 19),
    FruitData("Kiwi Fruit", 28),
    FruitData("Mangoes", 30),
    FruitData("Grapes", 16)
]

Color = np.array

fruitColors = {
    "Avocados": Color([0.0, 0.5, 0.0]),
    "Oranges": Color([1.0, 0.5, 0.0]),
    "Bananas": Color([1.0, 1.0, 0.0]),
    "Kiwi Fruit": Color([0.6, 0.3, 0.0]),
    "Mangoes": Color([1.0, 0.95, 0.0]),
    "Grapes": Color([0.5, 0.0, 0.5])
}

def drawAxes():
    gl.glColor3f(0.0, 0.0, 0.0)
    gl.glBegin(gl.GL_LINES)
    gl.glVertex2f(0.1, 0.1)
    gl.glVertex2f(0.1, 0.9)
    gl.glEnd()
    
    gl.glBegin(gl.GL_LINES)
    gl.glVertex2f(0.1, 0.1)
    gl.glVertex2f(0.9, 0.1)
    gl.glEnd()
    
    gl.glColor3f(1.0, 0.0, 0.0)
    labelY = 0.1
    for i in range(0, 56, 13):
        gl.glRasterPos2f(0.05, labelY)
        label = str(i)
        for r in label:
            glut.glutBitmapCharacter(glut.GLUT_BITMAP_HELVETICA_10, ord(r))
        labelY += 0.2

def drawBars():
    barWidth = 0.08
    startX = 0.18
    startY = 0.1
    maxHeight = 0.8
    for i in range(len(fruitData)):
        color = fruitColors[fruitData[i].name]
        gl.glColor3fv(color)
        barHeight = fruitData[i].quantity / 50.0
        gl.glBegin(gl.GL_QUADS)
        gl.glVertex2f(startX + i * 0.10, startY)
        gl.glVertex2f(startX + i * 0.10, startY + barHeight * maxHeight)
        gl.glVertex2f(startX + i * 0.10 + barWidth, startY + barHeight * maxHeight)
        gl.glVertex2f(startX + i * 0.10 + barWidth, startY)
        gl.glEnd()

def drawLabels():
    labelX = 0.2
    labelY = 0.05
    for i in range(len(fruitData)):
        gl.glColor3f(0.0, 0.0, 0.0)
        gl.glRasterPos2f(labelX + i * 0.10, labelY)
        for c in fruitData[i].name:
            glut.glutBitmapCharacter(glut.GLUT_BITMAP_TIMES_ROMAN_10, ord(c))

def display():
    gl.glClear(gl.GL_COLOR_BUFFER_BIT)
    drawAxes()
    drawBars()
    drawLabels()
    gl.glFlush()

def main():
    glut.glutInit()
    glut.glutInitDisplayMode(glut.GLUT_SINGLE | glut.GLUT_RGB)
    glut.glutInitWindowSize(800, 600)
    glut.glutCreateWindow(b"Fruit Preference Bar Chart")
    glut.glutDisplayFunc(display)
    gl.glClearColor(1.0, 1.0, 1.0, 1.0)
    gl.glOrtho(0.0, 1.0, 0.0, 1.0, -1.0, 1.0)
    glut.glutMainLoop()

if __name__ == "__main__":
    main()

#ran in a virtual environment

File: bar2.cpp
#include <GL/glut.h>
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <array> // Include array header for using std::array
using namespace std;

struct FruitData {
    string name;
    int quantity;
};

vector<FruitData> fruitData = {
    {"Avocados", 36},
    {"Oranges", 41},
    {"Bananas", 19},
    {"Kiwi Fruit", 28},
    {"Mangoes", 30},
    {"Grapes", 16}
};

// Defining a type alias for the color array
using Color = array<float, 3>;

// Define a map to store fruit colors
map<string, Color> fruitColors = {
    {"Avocados", {0.0f, 0.5f, 0.0f}},   // slightly dark-green for Avocados
    {"Oranges", {1.0f, 0.5f, 0.0f}},     // Orange for Oranges
    {"Bananas", {1.0f, 1.0f, 0.0f}},     // Yellow for Bananas
    {"Kiwi Fruit", {0.6f, 0.3f, 0.0f}},  // Brown green for Kiwi Fruit
    {"Mangoes", {1.0f, 0.95f, 0.0f}},    // Mango color for Mangoes
    {"Grapes", {0.5f, 0.0f, 0.5f}}       // Purple for Grapes
};

void drawAxes() {
    // Draw horizontal axis
    glColor3f(0.0, 0.0, 0.0);
    glBegin(GL_LINES);
    glVertex2f(0.1, 0.1);//origin
    glVertex2f(0.1, 0.9);//y-max
    glEnd();

    // Draw vertical axis
    glBegin(GL_LINES);
    glVertex2f(0.1, 0.1);//x-min
    glVertex2f(0.9, 0.1); //y-max
    glEnd();

    glColor3f(1.0, 0.0, 0.0 );
    float labelY = 0.1;
    for (int i = 0; i <=55; i+=13){ //y-axis label  
        glRasterPos2f(0.05, labelY);
        string label = to_string(i);
        for (char r:label){
            glutBitmapCharacter(GLUT_BITMAP_HELVETICA_10, r);
        }
        labelY += 0.2;
    }

}

void drawBars() {
    float barWidth = 0.08; // width of the bars
    float startX = 0.18;
    float startY = 0.1;
    float maxHeight = 0.8;

    for (size_t i = 0; i < fruitData.size(); ++i) {
        // Get color based on fruit name
        Color color = fruitColors[fruitData[i].name];

        glColor3fv(color.data()); // Set color for the bar
        float barHeight = (float)fruitData[i].quantity / 50.0; // Scale the height of the bars
        glBegin(GL_QUADS);
        glVertex2f(startX + i * 0.10, startY);
        glVertex2f(startX + i * 0.10, startY + barHeight * maxHeight);
        glVertex2f(startX + i * 0.10 + barWidth, startY + barHeight * maxHeight);
        glVertex2f(startX + i * 0.10 + barWidth, startY);
        
        //the 0.10 represents the bar width
        
        glEnd();
    }
}

void drawLabels() { //the labels for the bars for every fruit distro
    float labelX = 0.2;
    float labelY = 0.05;

    for (size_t i = 0; i < fruitData.size(); ++i) {
        glColor3f(0.0, 0.0, 0.0); // Black text
        glRasterPos2f(labelX + i * 0.10, labelY); //
        for (char c : fruitData[i].name) {
            // 
            
            glutBitmapCharacter(GLUT_BITMAP_TIMES_ROMAN_10, c);
        }
    }
}

void display() {
    glClear(GL_COLOR_BUFFER_BIT);
    drawAxes();
    drawBars();
    drawLabels();
    glFlush();
}

int main(int argc, char** argv) {
    glutInit(&argc, argv);
    glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB);
    glutInitWindowSize(800, 600);
    glutCreateWindow("Fruit Preference Bar Chart");

    glutDisplayFunc(display);
    glClearColor(1.0, 1.0, 1.0, 1.0);
    gluOrtho2D(0.0, 1.0, 0.0, 1.0);
    glutMainLoop();
    return 0;
}
